dspace-api/src/main/java/org/dspace/google/GoogleAsyncEventListener.java

```diff
@@ -0,0 +1,236 @@
+                            for (int x = 1; x <= context.getEvents().size(); x++) {
+                                log.error("    Context Event " + x + ": " + context.getEvents().get(x));
+                            }
+                        }
+                    } else {
+                        log.error("UsageEvent has no Context object");
+                    }
+                }
+            }
+        }
+    }
+
+    private String getObjectName(UsageEvent ue) {
+        try {
+            if (ue.getObject().getType() == Constants.BITSTREAM) {
+                // For a bitstream download we really want to know the title of the owning item
+                // rather than the bitstream name.
+                return ContentServiceFactory.getInstance().getDSpaceObjectService(ue.getObject())
+                        .getParentObject(ue.getContext(), ue.getObject()).getName();
+            } else {
+                return ue.getObject().getName();
+            }
+        } catch (SQLException e) {
+            // This shouldn't merit interrupting the user's transaction so log the error and continue.
+            log.error("Error in Google Analytics recording - can't determine ParentObjectName for bitstream " +
+                    ue.getObject().getID(), e);
+        }
+
+        return null;
+
+    }
+
+    private String getIPAddress(HttpServletRequest request) {
+        String clientIP = request.getRemoteAddr();
+        if (configurationService.getBooleanProperty("useProxies", false) &&
+                request.getHeader("X-Forwarded-For") != null) {
+            /* This header is a comma delimited list */
+            for (String xfip : request.getHeader("X-Forwarded-For").split(",")) {
+                /* proxy itself will sometime populate this header with the same value in
+                    remote address. ordering in spec is vague, we'll just take the last
+                    not equal to the proxy
+                */
+                if (!request.getHeader("X-Forwarded-For").contains(clientIP)) {
+                    clientIP = xfip.trim();
+                }
+            }
+        }
+
+        return clientIP;
+    }
+
+    @PreDestroy
+    public void destroy() throws InterruptedException {
+        destroyed = true;
+        if (StringUtils.isNotEmpty(analyticsKey)) {
+            future.cancel(true);
+            executor.shutdown();
+            executor.awaitTermination(1, TimeUnit.SECONDS);
+        }
+    }
+
+    private static class GoogleAnalyticsTask implements Runnable {
+        public void run() {
+            while (!destroyed) {
+                try {
+                    boolean sleep = false;
+                    StringBuilder request = null;
+                    List<GoogleAnalyticsEvent> events = new ArrayList<>();
+                    Iterator iterator = buffer.iterator();
+                    for (int x = 0; x < 20 && iterator.hasNext(); x++) {
```

Comment:
I would find it useful to note here, like you did in the PR introduction, that 20 is the max set by the GA API. Just so that whoever comes along realises that it isn't arbitrarily selected.

Reply:
I agree that a comment regarding the 20 would be useful here.